### PR TITLE
Add editorconfig-checker to atmos cli

### PR DESCRIPTION
**Pull request by samtholiya**
issue: https://linear.app/cloudposse/issue/DEV-2836/implement-atmos-validate-editorconfig

## what
* Added functionality to atmos validate to validate against the .editorconfig file if it is defined.
* Integrated the editorconfig-checker library to perform the validation.
* Modified the validation process to support OPA/JSON schema validation alongside the new .editorconfig validation.

## why

* The change ensures that the .editorconfig file is validated, which helps maintain consistency in code formatting across teams.
* By incorporating editorconfig-checker, the validation process becomes more comprehensive, ensuring that both OPA/JSON schemas and editor configuration are correct.
* This update was made to improve development workflow by catching formatting issues early during the validation process.

## references

* https://github.com/editorconfig-checker/editorconfig-checker
* https://linear.app/cloudposse/issue/DEV-2836/implement-atmos-validate-editorconfig

### CODE REVIEW

File: cmd/editor_config.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	er "github.com/editorconfig-checker/editorconfig-checker/v3/pkg/error"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/files"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/utils"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/validation"
+	"github.com/spf13/cobra"
+)
+
+var (
+	editorConfigVersion   = "v3.0.3"
+	defaultConfigFilePath = ".ecrc"
```
Comment (osterman) [inline cmd/editor_config.go:19]: Let's use the more conventional filename of .editorconfig

File: cmd/editor_config.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	currentConfig         *config.Config
+	cliConfig             config.Config
+	configFilePath        string
+	tmpExclude            string
+)
+
+var editorConfigCmd *cobra.Command = &cobra.Command{
+	Use:   "editorconfig-checker",
+	Short: "EditorConfig Checker CLI",
```
Comment (osterman) [inline cmd/editor_config.go:28]:
```suggestion
	Short: "Validate all files against the EditorConfig",
```

File: cmd/editor_config.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	Long:  "A command-line tool to check your files against the EditorConfig rules",
```
Comment (osterman) [inline cmd/editor_config.go:29]:
```suggestion
	Long:  "Validate all files against the project's EditorConfig rules",
```

File: cmd/editor_config.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+// addPersistentFlags adds flags to the root command
+func addPersistentFlags(cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVar(&configFilePath, "config", "", "Path to the configuration file")
+	cmd.PersistentFlags().StringVar(&tmpExclude, "exclude", "", "Regex to exclude files from checking")
+	cmd.PersistentFlags().BoolVar(&cliConfig.IgnoreDefaults, "ignore-defaults", false, "Ignore default excludes")
+	cmd.PersistentFlags().BoolVar(&cliConfig.DryRun, "dry-run", false, "Show which files would be checked")
+	cmd.PersistentFlags().BoolVar(&cliConfig.ShowVersion, "version", false, "Print the version number")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Help, "help", false, "Print help information")
+	cmd.PersistentFlags().StringVar(&cliConfig.Format, "format", "default", "Specify the output format: default, gcc")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Verbose, "verbose", false, "Print debugging information")
```
Comment (osterman) [inline cmd/editor_config.go:138]: This should probably not be a new flag, but be synchronized with the atmos log level

File: cmd/editor_config.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	return false
+}
+
+// addPersistentFlags adds flags to the root command
+func addPersistentFlags(cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVar(&configFilePath, "config", "", "Path to the configuration file")
+	cmd.PersistentFlags().StringVar(&tmpExclude, "exclude", "", "Regex to exclude files from checking")
+	cmd.PersistentFlags().BoolVar(&cliConfig.IgnoreDefaults, "ignore-defaults", false, "Ignore default excludes")
+	cmd.PersistentFlags().BoolVar(&cliConfig.DryRun, "dry-run", false, "Show which files would be checked")
+	cmd.PersistentFlags().BoolVar(&cliConfig.ShowVersion, "version", false, "Print the version number")
```
Comment (osterman) [inline cmd/editor_config.go:135]: This flag doesn't make sense. Drop it.

File: cmd/editor_config.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if config.Help {
+		config.Logger.Output("USAGE:")
+		return true
+	}
+	return false
+}
+
+// addPersistentFlags adds flags to the root command
+func addPersistentFlags(cmd *cobra.Command) {
```
Comment (osterman) [inline cmd/editor_config.go:130]: All these flags should be supported with defaults in atmos.yaml. See other comment.

```
validate:
  editorconfig:
    ...
```

File: examples/demo-stacks/.ecrc [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# EditorConfig for Terraform repository
```
Comment (osterman) [inline examples/demo-stacks/.ecrc:1]: Rename to the more conventional filename of .editorconfig

File: tests/test_cases.yaml [hunk 8]
```diff
@@ -136,7 +136,21 @@ tests:
       stderr:
         - 'unknown command "non-existent" for "atmos"'
       exit_code: 1
-
+  - name: atmos validate editorconfig-check
```
Comment (osterman) [inline tests/test_cases.yaml:139]:
```suggestion
  - name: atmos validate editorconfig
```

File: cmd/editor_config.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	editorConfigVersion   = "v3.0.3"
+	defaultConfigFilePath = ".ecrc"
+	currentConfig         *config.Config
+	cliConfig             config.Config
+	configFilePath        string
+	tmpExclude            string
+)
+
+var editorConfigCmd *cobra.Command = &cobra.Command{
+	Use:   "editorconfig-checker",
```
Comment (osterman) [inline cmd/editor_config.go:27]:
```suggestion
	Use:   "editorconfig",
```

File: cmd/editor_config.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		os.Exit(0)
+	}
+
+	errors := validation.ProcessValidation(filePaths, config)
+	errorCount := er.GetErrorCount(errors)
+
+	if errorCount != 0 {
+		er.PrintErrors(errors, config)
+		u.LogError(atmosConfig, fmt.Errorf("\n%d errors found", errorCount))
```
Comment (osterman) [inline cmd/editor_config.go:97]: I think we have a function that logs error and exits

File: cmd/editor_config.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		os.Exit(1)
+	}
+
+	u.LogTrace(atmosConfig, fmt.Sprintf("%d files checked", len(filePaths)))
```
Comment (osterman) [inline cmd/editor_config.go:101]:
```suggestion
	u.LogDebug(atmosConfig, fmt.Sprintf("%d files checked", len(filePaths)))
```

File: cmd/editor_config.go (hunk 5 above)
Comment (samtholiya) [reply cmd/editor_config.go:136]: I thought we were refering to editorconfig-checker. This would help our customer with better documentation and understanding of how our editorconfig-checker works.

Does this makes sense?

File: cmd/editor_config.go (hunk 4 above)
Comment (samtholiya) [reply cmd/editor_config.go:139]: https://github.com/cloudposse/atmos/pull/896#discussion_r1898901575

File: cmd/editor_config.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	var err error
+	currentConfig, err = config.NewConfig(configFilePath)
+	if err != nil {
+		u.LogError(atmosConfig, err)
+		os.Exit(1)
+	}
+
+	_ = currentConfig.Parse()
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/editor_config.go:60]: We would not be fixing this one because:
It is as per the code of editorconfig. I tried to understand why it is but i guess it is not worth looking into why they wrote it that way. Let us just keep it as is and update it if we find any issue. Currently i found it working perfect. I updated the tests as well

File: cmd/editor_config.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+		initializeConfig()
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		runMainLogic()
+	},
+}
+
+// initializeConfig breaks the initialization cycle by separating the config setup
+func initializeConfig() {
+	u.LogInfo(atmosConfig, fmt.Sprintf("EditorConfig Checker CLI Version: %s", editorConfigVersion))
```
Comment (osterman) [inline cmd/editor_config.go:41]:
```suggestion
```

File: cmd/editor_config.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/config"
+	er "github.com/editorconfig-checker/editorconfig-checker/v3/pkg/error"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/files"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/utils"
+	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/validation"
+	"github.com/spf13/cobra"
+)
+
+var (
+	editorConfigVersion   = "v3.0.3"
```
Comment (osterman) [inline cmd/editor_config.go:18]:
```suggestion
```

File: cmd/editor_config.go (hunk 5 above)
Comment (osterman) [reply cmd/editor_config.go:138]: Since we are using it as a module, we should document it and maintain it like a feature of atmos.

Comment (osterman) [reply cmd/editor_config.go:138]: The flag I was referring to was the version flag. It doesn't make sense that a sub command natively implemented into atmos has a separate version. Atmos depends on hundreds of modules, each with their own version. This is simply another module.

File: examples/quick-start-simple/atmos.yaml [hunk 15]
```diff
@@ -17,5 +17,4 @@ stacks:
   name_pattern: "{stage}"
 
 logs:
-  file: "/dev/stderr"
```
Comment (samtholiya) [inline examples/quick-start-simple/atmos.yaml:20]: I suggest removing this from quick start simple 

Especially because this config causes all the logs to direct towards error and a quick start guide should not do that.

Of course we can document and give a seperate example (which i found many already in the repo) for this behavior.

File: cmd/editor_config.go (hunk 13 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+	cmd.PersistentFlags().StringVar(&cliConfig.Format, "format", "default", "Specify the output format: default, gcc")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Verbose, "verbose", false, "Print debugging information")
+	cmd.PersistentFlags().BoolVar(&cliConfig.NoColor, "no-color", false, "Don't print colors")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.TrimTrailingWhitespace, "disable-trim-trailing-whitespace", false, "Disable trailing whitespace check")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.EndOfLine, "disable-end-of-line", false, "Disable end-of-line check")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.InsertFinalNewline, "disable-insert-final-newline", false, "Disable final newline check")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.Indentation, "disable-indentation", false, "Disable indentation check")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.IndentSize, "disable-indent-size", false, "Disable indent size check")
+	cmd.PersistentFlags().BoolVar(&cliConfig.Disable.MaxLineLength, "disable-max-line-length", false, "Disable max line length check")
+}
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 4 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 1 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 2 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 3 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 5 above)
Comment (samtholiya) [reply cmd/editor_config.go:173]: Done

File: cmd/editor_config.go (hunk 14 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 11 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 10 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: cmd/editor_config.go (hunk 9 above)
Comment (samtholiya) [reply cmd/editor_config.go]: Done

File: examples/demo-stacks/.ecrc (hunk 7 above)
Comment (samtholiya) [reply examples/demo-stacks/.ecrc]: Done

File: cmd/editor_config.go (hunk 6 above)
Comment (samtholiya) [reply cmd/editor_config.go:166]: Done

**Comment by osterman** [comment]
Add the docs as a subpage to: https://atmos.tools/cli/commands/validate/usage

File: tests/test_cases.yaml [hunk 17]
```diff
@@ -136,7 +136,21 @@ tests:
...
+    enabled: true
+    description: "Ensure atmos CLI validates based on the .ecrc file."
+    workdir: "../examples/quick-start-simple/"
+    command: "atmos"
+    args:
+      - "validate"
+      - "editorconfig"
+    expect:
+      stdout:
+        - "EditorConfig Checker CLI Version: v3.0.3"
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply tests/test_cases.yaml]: This is out dated. It was already removed. Not sure why you got the old diff

File: .github/workflows/test.yml [hunk 18]
```diff
@@ -86,6 +86,11 @@ jobs:
     timeout-minutes: 15
     runs-on: ${{ matrix.flavor.os }}
     steps:
+      - name: Set Git Preferences for windows
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply .github/workflows/test.yml:89]: Will add the comment. It is related to how git clones the repo.

File: examples/quick-start-simple/atmos.yaml (hunk 15 above)
Comment (samtholiya) [reply examples/quick-start-simple/atmos.yaml:20]: Based on the code in the following two places;
https://github.com/cloudposse/atmos/blob/main/pkg/logger/logger.go#L61-L98
https://github.com/cloudposse/atmos/blob/main/pkg/utils/log_utils.go#L106-L143

The log emitted is not based on the log level if we set the log file to stderr or stdout.
This caused my code with `u.LogInfo` to be written in stderr terminal.

I assumed this is some required behaviour given that the file path were prefixed with `/dev`
although in hindsight it seems why we would require this code at all?

Log level is all we should be needing.

Just FYI, I think due to above behavior we are using empty atmosConfig here; https://github.com/cloudposse/atmos/blob/main/cmd/cmd_utils.go#L170-L199

Though this would work in cases where user did not set a file for log, it won't if he sets a file for log because using empty config should not be great.

Comment (samtholiya) [reply examples/quick-start-simple/atmos.yaml:20]: @aknysh 
Thank you for the explanation regarding logs.file and logs.level. I agree that both are necessary for flexibility. However, my concern is about the specific behavior when the log destination (logs.file) is set to /dev/stderr or /dev/stdout.

Currently, based on the code:

Logs are written to stderr or stdout directly, regardless of the log level set in logs.level.
This behavior causes logs (e.g., Info) to appear in stderr, which might not align with user expectations.
Additionally:

In cmd_utils.go [L170-L199](https://github.com/cloudposse/atmos/blob/main/cmd/cmd_utils.go#L170-L199), the use of an empty atmosConfig works fine when no file is specified for logs.file. However, if a user configures a specific file for logs.file, this logic may not correctly redirect logs to the specified file.
Could you confirm if this is the intended behavior, or would it make sense to update the implementation to ensure that:

Logs respect both logs.file and logs.level settings, even when writing to stderr or stdout.
cmd_utils.go handles configurations with logs.file more robustly?

How to reproduce the issue:

Run `atmos validate editorconfig > stdout.log 2> stderr.log` With the log file set to `/dev/stderr` in `atmos.yaml`

What will happen?
You will see even the logs that are info/output of the library are published to stderr.log

This is the reason the test case fails because all logs should not go to `stderr.log` info logs should go `stdout.log`

Comment (osterman) [reply examples/quick-start-simple/atmos.yaml:20]: Thanks for the write up! If this is the current behavior, then it is indeed wrong. 

For example, if `logs.file` is set to `/dev/stderr` and `logs.level` is `Debug`, I would only expect to see {Info, Warn, Error, and Debug} messages.

If `logs.file` is set to `/dev/stderr` and `logs.level` is `Info`, I would only expect to see {Info} messages.

If `logs.file` is set to `/dev/stderr` and `logs.level` is `Trace`, I would expect to see every single log level message.

File: atmos.yaml [hunk 19]
```diff
@@ -91,6 +91,44 @@ logs:
   # Can also be set using 'ATMOS_LOGS_LEVEL' ENV var, or '--logs-level' command-line argument
   level: Info
 
+validate:
+  # The configuration settings for the editorconfig-checker.
+
+  editorconfig:
+    # A comma-separated list of file paths or patterns to exclude from checks.
+    exclude: ""
```
Comment (osterman) [inline atmos.yaml:99]: This is wrong. It should be a list. It's also a list in their examples.

```
"Exclude": ["testfiles", "\\.md$"],
```

Comment (osterman) [inline atmos.yaml:99]:
```suggestion
    exclude: []
```

File: atmos.yaml [hunk 20]
```diff
@@ -91,6 +91,44 @@ logs:
   # Can also be set using 'ATMOS_LOGS_LEVEL' ENV var, or '--logs-level' command-line argument
   level: Info
 
+validate:
+  # The configuration settings for the editorconfig-checker.
+
+  editorconfig:
+    # A comma-separated list of file paths or patterns to exclude from checks.
```
Comment (osterman) [inline atmos.yaml:98]:
```suggestion
    # A list of file paths or patterns to exclude from checks.
```

File: atmos.yaml [hunk 21]
```diff
@@ -91,6 +91,44 @@ logs:
...
+    ignore_defaults: false
+
+    # Runs the checker without making any changes or producing output, useful for testing configuration.
+    dry_run: false
+
+    # Specifies the output format. Options: "default", "json".
+    format: "default"
+
+    # Disables colored output in the terminal.
+    no_color: false
```
Comment (osterman) [inline atmos.yaml:111]: Let's use atmos conventions of positive booleans, so we don't use double negatives.

```suggestion
    # Toggles colored output in the terminal.
    color: true
```

File: atmos.yaml [hunk 22]
```diff
@@ -91,6 +91,44 @@ logs:
...
+
+    disable:
```
Comment (osterman) [inline atmos.yaml:113]: Can we just make this at the same level of `validate.editorconfig`? I don't think it needs to be nested. If it needs to be nested, then we need a different section name than `disabled`. That said, strongly prefer we eliminate nesting.

```suggestion
    disable:
```

File: pkg/schema/schema.go [hunk 23]
```diff
@@ -30,13 +32,39 @@ type AtmosConfiguration struct {
...
+	NoColor        bool           `yaml:"no_color,omitempty" json:"no_color,omitempty" mapstructure:"no_color"`
+	Disable        DisabledChecks `yaml:"disable,omitempty" json:"disable,omitempty" mapstructure:"disable"`
+
+	ConfigFilePath string `yaml:"config_file_path,omitempty" json:"config_file_path,omitempty" mapstructure:"config_file_path"`
+	Exclude        string `yaml:"exclude,omitempty" json:"exclude,omitempty" mapstructure:"exclude"`
+	Init           bool   `yaml:"init,omitempty" json:"init,omitempty" mapstructure:"init"`
+}
+
+// DisabledChecks is a Struct which represents disabled checks
+type DisabledChecks struct {
```
Comment (osterman) [inline pkg/schema/schema.go:59]:
```suggestion
// ConfigOptions represents configurable feature flags
type ConfigOptions struct {
}
```

File: tests/test_cases.yaml [hunk 24]
```diff
@@ -136,7 +136,20 @@ tests:
       stderr:
         - 'unknown command "non-existent" for "atmos"'
       exit_code: 1
-
+  - name: atmos validate editorconfig
```
Comment (osterman) [inline tests/test_cases.yaml:139]: Great

Comment (osterman) [reply tests/test_cases.yaml:139]: Maybe add one check that verifies the negative.

File: examples/quick-start-simple/atmos.yaml (hunk 15 above)
Comment (samtholiya) [reply examples/quick-start-simple/atmos.yaml:20]: Resolved this confusion with @aknysh by having a screen share.
Understood that this is working as intended. My confusion regarding the terminal error and terminal output was solved by him and came to conclusion that this is intended behavior.

File: pkg/schema/schema.go (hunk 23 above)
Comment (samtholiya) [reply pkg/schema/schema.go:59]: Naming it `ConfigOptions` Feels a bit misleading. I understand `DisableChecks` is also unique and maybe not the perfect answer but considering that this is used by `EditorConfig` only. `ConfigOptions` feel a bit generic for this.

File: tests/test_cases.yaml (hunk 24 above)
Comment (samtholiya) [reply tests/test_cases.yaml:139]: Added one :)

File: pkg/schema/schema.go (hunk 23 above)
Comment (osterman) [reply pkg/schema/schema.go:59]: I agree that it's too generic for this file. I thought we were in a package specific to editor config. How about calling it EditorConfigOptions or something along those lines.

File: tests/test_cases.yaml [hunk 25]
```diff
@@ -136,6 +136,32 @@ tests:
...
+    expect:
+      stdout:
+        - "No errors found"
+      stderr:
+        - "^$"
+      exit_code: 0
+  - name: atmos validate editorconfig error
+    enabled: true
+    description: "Ensure atmos CLI gives error during validation based on editorconfig"
+    workdir: "../examples/quick-start-advanced/"
```
Comment (osterman) [inline tests/test_cases.yaml:156]: Heh, it hurts that we are using a quick star for a negative case, when instead we should fix the quick start. Instead let's use example/tests, which is expected to have errors for testing.

File: pkg/schema/schema.go (hunk 23 above)
Comment (samtholiya) [reply pkg/schema/schema.go:59]: Sugestion from chatgpt based on the structure we have.

ValidationOverrides: Suggests these are overrides to default validation rules.
SkippedValidations: Indicates validations that will not be performed.
IgnoredValidationRules: Makes it explicit that these are rules being ignored.
ValidationToggles: Implies that these are specific toggles for enabling/disabling validations.

Comment (samtholiya) [reply pkg/schema/schema.go:59]: I am not inclined towards `EditorConfigOptions ` because our heirarchy is 
AtmosConfiguration > Validate > EditorConfig > DisabledChecks

So `EditorConfigOptions` is a bit odd considering EditorConfig structs itself hold EditorConfig Options

Comment (osterman) [reply pkg/schema/schema.go:59]: Why can't these just be directly underneath the editor config? Why do we need a second level of nesting?

File: atmos.yaml (hunk 21 above)
Comment (samtholiya) [reply atmos.yaml:111]: Done.

File: pkg/schema/schema.go (hunk 23 above)
Comment (samtholiya) [reply pkg/schema/schema.go]: Done

File: atmos.yaml (hunk 19 above)
Comment (samtholiya) [reply atmos.yaml]: Done

File: atmos.yaml (hunk 20 above)
Comment (samtholiya) [reply atmos.yaml]: Done

File: atmos.yaml (hunk 19 above)
Comment (samtholiya) [reply atmos.yaml]: Done

File: tests/test_cases.yaml [hunk 26]
```diff
@@ -136,6 +136,32 @@ tests:
...
+    description: "Ensure atmos CLI gives error during validation based on editorconfig"
+    workdir: "../examples/tests/"
+    command: "atmos"
+    args:
+      - "validate"
+      - "editorconfig"
+    expect:
+      stderr:
+        - "18 errors found"
+      exit_code: 1
```
Comment (osterman) [inline tests/test_cases.yaml:164]: Test cases have moved:
- #918

**Comment by osterman** [comment]
- [x] add a new page under this https://atmos.tools/core-concepts/validate/ with documentation on how to configure editor config

File: tests/test_cases.yaml (hunk 26 above)
Comment (samtholiya) [reply tests/test_cases.yaml]: Done

File: cmd/editor_config.go [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+	Short: "Validate all files against the EditorConfig",
+	Long:  "Validate all files against the project's EditorConfig rules",
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		initializeConfig(cmd)
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		if cliConfig.Help {
+			cmd.Help()
+			os.Exit(0)
+		}
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/editor_config.go:40]: This part will change anyway so i am not worried about it as of now.

File: tests/test_cases.yaml
Comment (osterman) [inline tests/test_cases.yaml:1]: This file has moved

File: website/docs/core-concepts/validate/validate.mdx [hunk 28]
```diff
@@ -55,41 +59,41 @@ The command checks and validates the following:
...
-  be displayed similar to the following:
+defined in more than one stack manifest file, and the component configurations are different, an error message will
+be displayed similar to the following:
 
-  <Terminal title="atmos validate stacks">
-    ```console
-    The Atmos component 'vpc' in the stack 'plat-ue2-dev' is defined in more than one
-    top-level stack manifest file: orgs/acme/plat/dev/us-east-2-extras, orgs/acme/plat/dev/us-east-2.
+<Terminal title="atmos validate stacks">
+```console
```
Comment (osterman) [inline website/docs/core-concepts/validate/validate.mdx:66]: Changing the indention now means this is no longer associated with the preceeding bullet point.

File: .github/workflows/test.yml [hunk 29]
```diff
@@ -86,6 +86,14 @@ jobs:
...
     runs-on: ${{ matrix.flavor.os }}
     steps:
+      # Configures Git to ensure LF line endings on Windows to pass the `atmos validate editorconfig` check.
+      # The check expects LF endings, but on Windows, files might default to CRLF due to core.autocrlf.
+      # Since the validation runs in CI, we enforce consistent settings across all OS environments.
+      - name: Set Git Preferences for windows
+        if: matrix.flavor.target == 'windows'
+        run: |
+          git config --global core.autocrlf false
+          git config --global core.eol lf
```
Comment (osterman) [inline .github/workflows/test.yml:96]: Let's move this to the `.gitattributes` file instead, and remove it from here.

Comment (osterman) [reply .github/workflows/test.yml:96]: https://github.com/cloudposse/atmos/blob/main/.gitattributes

```
* text=auto
```

Comment (osterman) [reply .github/workflows/test.yml:96]: @mcalhoun is going to open a PR to address the root problem.

File: tests/test_cases.yaml
Comment (samtholiya) [reply tests/test_cases.yaml:1]: Done

File: website/src/components/Screengrabs/atmos-validate-editorconfig--help.html [hunk 30]
```diff
@@ -0,0 +1,389 @@
...
+
+.term a { color: inherit; text-decoration: underline; text-decoration-style: dashed; }
+.term a:hover { color: #2882F9 }
+
+@keyframes blink-animation {
+  to {
+    visibility: hidden;
+  }
+}
+
```
Comment (samtholiya) [inline website/src/components/Screengrabs/atmos-validate-editorconfig--help.html:32]: This is not the perfect version. I am trying to get the perfect versions like you got.

File: website/docs/cli/commands/validate/validate-editorconfig.mdx [hunk 31]
```diff
@@ -0,0 +1,57 @@
+---
+title: atmos validate editorconfig
+sidebar_label: editorconfig
+sidebar_class_name: command
+id: editorconfig
+description: Use this command to validate an Atmos component in a stack using JSON Schema and OPA policies.
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply website/docs/cli/commands/validate/validate-editorconfig.mdx:6]: Nope

File: cmd/editor_config.go [hunk 32]
```diff
@@ -0,0 +1,194 @@
+package cmd
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/editor_config.go:1]: Ohh yes ofcourse.
It should be `validate_editorconfig.go` to be precise, so using that.

File: cmd/validate_editorconfig.go [hunk 33]
```diff
@@ -0,0 +1,194 @@
...
+var editorConfigCmd *cobra.Command = &cobra.Command{
+	Use:   "editorconfig",
+	Short: "Validate all files against the EditorConfig",
+	Long:  "Validate all files against the project's EditorConfig rules",
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		initializeConfig(cmd)
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		if cliConfig.Help {
+			cmd.Help()
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/validate_editorconfig.go:38]: not required. We are going to change this soon anyway

File: cmd/validate_editorconfig.go [hunk 34]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	if initEditorConfig {
+		err := currentConfig.Save(version.Version)
+		if err != nil {
+			u.LogErrorAndExit(atmosConfig, err)
+		}
+	}
+
+	_ = currentConfig.Parse()
```
[the remark this answers is not in the document]
Comment (samtholiya) [reply cmd/validate_editorconfig.go:66]: we should not do this

File: .gitattributes [hunk 35]
```diff
@@ -16,3 +16,5 @@ website/src/components/Screengrabs/**/*.html linguist-generated=true binary
 *.eps binary
 *.ansi binary
 *.mp4 binary
+# Ensure all files use LF line endings
+* text eol=lf
```
Comment (osterman) [inline .gitattributes:20]: This resets what is already on lines 1-2.

Comment (osterman) [inline .gitattributes:20]:
```suggestion
```

File: .github/workflows/test.yml (hunk 29 above)
Comment (osterman) [inline .github/workflows/test.yml:96]: This should not be necessary with the `.gitattributes` fixes:
- #928 
- #929 
- #931 

```suggestion
```

File: tests/fixtures/.editorconfig
Comment (osterman) [inline tests/fixtures/.editorconfig:1]: This is the wrong location for the fixture.

image

I think for what you want, it would be in `scenarios/editorconfig/`

And you create a scenario there that tests the editor config. Otherwise, what we have now, will break if we ever add a new component to fixtures, or a new scenario - that changes the count. It's not a stable test.

File: tests/fixtures/scenarios/editorconfig/atmos.yaml [hunk 36]
```diff
@@ -0,0 +1,396 @@
...
+            description: Name of the component
+        flags:
+          - name: stack
+            shorthand: s
+            description: Name of the stack
+            required: true
+        env:
+          - key: ENV_VAR_1
+            value: ENV_VAR_1_value
+          - key: ENV_VAR_2
```
Comment (osterman) [inline tests/fixtures/scenarios/editorconfig/atmos.yaml:123]: Most of the configuration in this file, has nothing to do with testing editor config. Can we reduce it to something minimal for testing editor config?

File: tests/fixtures/scenarios/editorconfig/README.md [hunk 37]
```diff
@@ -0,0 +1,5 @@
+# Atmos Tests
+
+These configurations are used by the Atmos tests and CI to test/validate configurations and behaviors of Atmos. 
+Thus, they contain many deliberately broken examples and should not be used. 
+Instead, please see our [Quick Start](../quick-start) example.
```
Comment (osterman) [inline tests/fixtures/scenarios/editorconfig/README.md:5]:
```suggestion
# Atmos Editor Config Test Scenarios

This directory contains configurations specifically designed to test and validate the behavior of editor configurations when working with Atmos.

These files include intentionally broken examples and edge cases to help ensure that editors and tools handle Atmos configurations correctly. 

**Do not use these examples for actual deployments.**
```

File: tests/fixtures/scenarios/editorconfig/atmos.yaml (hunk 36 above)
Comment (samtholiya) [reply tests/fixtures/scenarios/editorconfig/atmos.yaml:123]: Done

File: tests/fixtures/scenarios/editorconfig/atmos.yaml
Comment (osterman) [inline tests/fixtures/scenarios/editorconfig/atmos.yaml:1]: Replace this with a minimal atmos configuration for testing editor config settings.

**Comment by aknysh** [comment]
@samtholiya please resolve the conflicts
